pulp_container/app/registry_api.py

```diff
@@ -743,14 +787,34 @@ def put(self, request, path, pk=None):
         try:
             tag.save()
         except IntegrityError:
-            pass
-        with repository.new_version() as new_version:
-            new_version.add_content(models.Manifest.objects.filter(digest=manifest.digest))
-            new_version.remove_content(models.Tag.objects.filter(name=tag.name))
-            new_version.add_content(
-                models.Tag.objects.filter(name=tag.name, tagged_manifest=manifest)
-            )
-        return ManifestResponse(manifest, path, request, status=201)
+            tag = models.Tag.objects.get(name=tag.name, tagged_manifest=manifest)
+
+        tags_to_remove = models.Tag.objects.filter(
+            pk__in=repository.latest_version().content.all(), name=tag
+        ).exclude(tagged_manifest=manifest)
+        task = enqueue_with_reservation(
+            add_and_remove,
+            [repository, blob],
+            kwargs={
+                "repository_pk": repository.pk,
+                "add_content_units": [tag.pk, manifest.pk],
+                "remove_content_units": tags_to_remove.values_list("pk"),
+            },
+        )
+
+        # Wait a small amount of time
+        for dummy in range(3):
+            time.sleep(1)
+            task = Task.objects.get(pk=task.id)
+            if task.state == "completed":
+                task.delete()
+                return ManifestResponse(manifest, path, request, status=201)
+            elif task.state in ["waiting", "running"]:
+                continue
+            else:
+                task.delete()
+                raise Exception("Failed.")
```

Comment:
I think we should delete the task before returning 429 regardless of it's state. Neither podman or docker retry during manifest put. Also they don't issue head before put so I don't see the reason of why to keep the task?

Reply:
OK. I will add an attempt to cancel it and then do another finished check and delete it.